### ISSUE TITLE
Regression: date format errors on "13" through "31"

### ISSUE THREAD

**Issue by aaronbeall** [handsontable/handsontable#5712]
### Description

Starting with `6.2.1` if you enter the values `13` through `31` in a date cell you end up with an error, while numbers like `10`-`12` and `32` and up are corrected to the year `2010`-`2012` and `2032`, etc. If you have `allowInvalid: false` enabled then the date will actually be corrected to the valid date, but the focus is captured in the cell. 

### Demo

https://jsfiddle.net/aqf0rmkc/1/

The date column is configured as:
```
      {
        type: 'date',
        dateFormat: 'MMM YYYY',
        correctFormat: true
      }
```

And the table has `allowInvalid: false`. 

The result is this:

![image](http://g.recordit.co/ltMVIiCHAW.gif)

Notice that with `6.2.0` entering any 2 digits like `12` and `13` and so on all result in the formatted date `Jan 2012`, `Jan 2013`, etc, but with `6.2.1` the values `13`-`31` result in formatting the date, but keeping the focus. If `allowInvalid: true` you end up with this:
![image](https://user-images.githubusercontent.com/2973900/51063772-6f83da80-15ca-11e9-9b60-f2b50c107bae.png)

### Your environment
* Handsontable version: 6.2.2
* Browser Name and version: Chrome 71.0.3578.98 (Official Build) (64-bit)
* Operating System: macOS Mojave 10.14.1

**Comment by aaronbeall** [comment]
> I run the `hot.validateCells()` (it's the same as reediting the cell). Then each first value of the date data

Yikes, that's problematic, but not actually the issue I am describing. In my case I shouldn't be setting initial data to a different format, it's just the case where a new value is entered into a cell.

> but with 6.2.1 the values 13-31 result in formatting the date, but keeping the focus

It sounds strange but *just* the numbers from `13` through `31` cause this behavior introduced with version `6.2.1`. The behavior (with the config mentioned above) is that after typing `13` and pressing ENTER it corrects the date to `Jan 13` but the input keeps focus. If `allowInvalid: true` then instead of correcting the date to `Jan 13` it just highlights the cell as an error.

Step by step:
1. Open the demo: https://jsfiddle.net/aqf0rmkc/1/
2. Click on the first "Registration Date" cell
3. Enter "12" and press ENTER -- note that it changes to "Jan 2012" and focuses the next cell, as expected
4. Enter "13" and press ENTER -- note that it changes to "Jan 2013" but the focus remains in the cell
![image](http://g.recordit.co/FJdqwEgJoF.gif)

Change `allowInvalid: false` to `allowInvalid: true` and repeat steps and after step 4 the cell remains "13" and is highlighted as an error:
![image](http://g.recordit.co/9pHVY815py.gif)

Change the version to `1.6.1` and the issue goes away, in other words "13" is corrected to "Jan 13" and there's no error, the same way that "12" is handled:
![image](http://g.recordit.co/pRmHTJh6yM.gif)

It's quite odd!

**Comment by AMBudnik** [comment]
You're right. I did not notice the selection. Not sure if that's one bug or two but surely something's wrong.

**Comment by wojciechczerniak** [comment]
I can't replicate exactly the range 13-31 here. In v5.0 it works up to 28 for example. In 6.2.2 all numbers are considered incorrect.

So I've tried to find who's responsible for that: 

`>> moment('13').toString()`
> Deprecation warning: value provided is not in a recognized RFC2822 or ISO format. moment construction falls back to js Date(), which is not reliable across all browsers and versions. Non RFC2822/ISO date formats are discouraged and will be removed in an upcoming major release. Please refer to http://momentjs.com/guides/#/warnings/js-date/ for more info.

Ok, `moment.js` won't parse that on its own. Let's try their fallback to js `Date()`:

`>> new Date('13')`
> Invalid Date

Nope, wrong turn. Back to `moment.js`? We have the `dateFormat` configured to `"MMM YYYY"`, let's try that:

`>> moment('13', 'MMM YYYY').toString()`
> "Tue Jan 01 2013 00:00:00 GMT+0100"

Great! We have the desired output. But why it doesn't work in Handsontable? Looks like we've introduced strict parsing flag for moment, third argument set to `true` in v6.2.1.

https://github.com/handsontable/handsontable/blob/c2ba2dc02e3eb883a58f976ef5944fa629dc981d/src/validators/dateValidator.js#L21

`>> moment('13', 'MMM YYYY', true).toString()`
> "Invalid date"

Yup, doesn't work again.

And if I change the format so it's parsed as years? `"YY"`

`>> moment('13', 'YY').toString()`
> "Tue Jan 01 2013 00:00:00 GMT+0100"

👍 

So is it really a regression? IMO was a bug and now it's fixed 🎉 But can we do better?

Sure, but how? Previous behavior sometimes was predictable:

`>> moment('13', 'MMM YYYY', false).toString()`
> "Tue Jan 01 2013 00:00:00 GMT+0100"

But...

`>> moment('4', 'MMM YYYY', false).toString()`
> "Thu Jan 01 0004 00:00:00 GMT+0130"

😲 

`>> moment('04', 'MMM YYYY', false).toString()`
> "Thu Jan 01 2004 00:00:00 GMT+0100"

Oh right, that makes sense... 🤔 

What we have to do here is to consider the difference between Pikaday and our DateValidator. Pikaday parses the value because it sets the strict flag only when you add this to config:

```
datePickerConfig: {
   formatStrict: true,
}
```

And our `strict` is hardcoded currently. Maybe we can read it from config and sync between those two. We should use `correctFormat` option for both?  Looks like it's ignored in some cases. But it's better to describe it in another issue as a feature / change. 

What do you think?

**Comment by aaronbeall** [comment]
Interesting, thanks for the walkthough analysis @wojciechczerniak . I now see that the inconsistent validation of `13-31` comes actually from the native `new Date()` part of the validator (`moment(new Date(valueToValidate)).isValid()`):
```
new Date("12") // Sat Dec 01 2001 00:00:00 GMT-0500 (Eastern Standard Time)
new Date("13") // Invalid Date
new Date("31") // Invalid Date
new Date("32") // Thu Jan 01 2032 00:00:00 GMT-0500 (Eastern Standard Time)
```
So the `dateValidator` accepts `12` but not `13` (as in the demo) even though both would be rejected by `moment(s, "MMM YYYY", true)`. Then the Pikaday formatter does something slightly different because it isn't strict like the validator.

> What do you think?

Yeah, I'm not sure? In my opinion the date _input_ with `correctFormat` should parse out almost any date format, not strictly the `dateFormat` of the cell... that's what I thought the purpose of `correctFormat` was. However, I don't think there's anything really worth fixing related to entering 2 digits as described inn this bug. Now that I understand what's going on, it's just asking for trouble to enter 2 digits in a `MMM YYYY` formatted cell. :)
